### ISSUE TITLE
IsADirectoryError when starting server

### ISSUE THREAD

**Issue by mparkms** [rr-/szurubooru#284]
Got through the setup process, but when starting the app using docker-compose, I see in the logs that the server always exits with the following error: 

```
server_1         | Traceback (most recent call last):
server_1         |   File "/usr/bin/alembic", line 11, in <module>
server_1         |     load_entry_point('alembic==1.2.1', 'console_scripts', 'alembic')()
server_1         |   File "/usr/lib/python3.7/site-packages/alembic/config.py", line 573, in main
server_1         |     CommandLine(prog=prog).main(argv=argv)
server_1         |   File "/usr/lib/python3.7/site-packages/alembic/config.py", line 567, in main
server_1         |     self.run_cmd(cfg, options)
server_1         |   File "/usr/lib/python3.7/site-packages/alembic/config.py", line 547, in run_cmd
server_1         |     **dict((k, getattr(options, k, None)) for k in kwarg)
server_1         |   File "/usr/lib/python3.7/site-packages/alembic/command.py", line 298, in upgrade
server_1         |     script.run_env()
server_1         |   File "/usr/lib/python3.7/site-packages/alembic/script/base.py", line 489, in run_env
server_1         |     util.load_python_file(self.dir, "env.py")
server_1         |   File "/usr/lib/python3.7/site-packages/alembic/util/pyfiles.py", line 98, in load_python_file
server_1         |     module = load_module_py(module_id, path)
server_1         |   File "/usr/lib/python3.7/site-packages/alembic/util/compat.py", line 173, in load_module_py
server_1         |     spec.loader.exec_module(module)
server_1         |   File "<frozen importlib._bootstrap_external>", line 728, in exec_module
server_1         |   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
server_1         |   File "szurubooru/migrations/env.py", line 12, in <module>
server_1         |     import szurubooru.model.base
server_1         |   File "/opt/app/szurubooru/migrations/../../szurubooru/model/__init__.py", line 2, in <module>
server_1         |     from szurubooru.model.user import User, UserToken
server_1         |   File "/opt/app/szurubooru/migrations/../../szurubooru/model/user.py", line 3, in <module>
server_1         |     from szurubooru.model.post import Post, PostScore, PostFavorite
server_1         |   File "/opt/app/szurubooru/migrations/../../szurubooru/model/post.py", line 4, in <module>
server_1         |     from szurubooru.model.comment import Comment
server_1         |   File "/opt/app/szurubooru/migrations/../../szurubooru/model/comment.py", line 2, in <module>
server_1         |     from szurubooru.db import get_session
server_1         |   File "/opt/app/szurubooru/migrations/../../szurubooru/db.py", line 5, in <module>
server_1         |     from szurubooru import config
server_1         |   File "/opt/app/szurubooru/migrations/../../szurubooru/config.py", line 64, in <module>
server_1         |     config = _read_config()  # pylint: disable=invalid-name
server_1         |   File "/opt/app/szurubooru/migrations/../../szurubooru/config.py", line 58, in _read_config
server_1         |     ret = _merge(ret, _file_config('config.yaml'))
server_1         |   File "/opt/app/szurubooru/migrations/../../szurubooru/config.py", line 51, in _file_config
server_1         |     with open(filename) as handle:
server_1         | IsADirectoryError: [Errno 21] Is a directory: 'config.yaml'
szurubooru_server_1 exited with code 1
```

config.yaml is, of course (or at least should be) a file, not a directory. I've copied config.yaml.dist as config.yaml before starting.

**Comment by sgsunder** [comment]
Are you sure you copied it to the right place? If you're using the provided `docker-compose.yml` from cloning the repository, it should be in `./server/config.yaml`. Can you check to see if that file exists in that location and is not a directory?

If you don't put a file in that location on first startup, Docker Compose automatically creates a directory in that location, then the server will error out when trying to read it as a file.

**Comment by mparkms** [comment]
Yes, config.yaml is in the right place as you can see: 
![image](https://user-images.githubusercontent.com/9259787/66448133-72024c00-ea05-11e9-91a1-c31cadc9d327.png)

I've tried restarting my Docker instance and uninstalling and reinstalling Docker altogether, but the error still persists.

**Comment by sgsunder** [comment]
Hmm, looking at your directory format, it looks like your using something like cygwin on Windows. Unfortunately, I have really no experience in running Docker outside of a Linux environment so I can't really diagnose why it isn't working for you

I found this issue https://github.com/docker/compose/issues/6242 which might be of some help.

**Comment by mparkms** [comment]
The first suggestion in that link (to use Shared Drives) fixed this problem. The directory I was in was not being shared with the Docker machine. 

However, now I'm running into a different issue, where the connection to ElasticSearch is failing. From the logs it seems like it's trying to connect before ElasticSearch has finished initializing and set up the endpoint at port 9200. See below:

```
server_1         | Traceback (most recent call last):
server_1         |   File "/usr/lib/python3.7/site-packages/urllib3/connection.py", line 157, in _new_conn
server_1         |     (self._dns_host, self.port), self.timeout, **extra_kw
server_1         |   File "/usr/lib/python3.7/site-packages/urllib3/util/connection.py", line 84, in create_connection
server_1         |     raise err
server_1         |   File "/usr/lib/python3.7/site-packages/urllib3/util/connection.py", line 74, in create_connection
server_1         |     sock.connect(sa)
server_1         | ConnectionRefusedError: [Errno 111] Connection refused
server_1         |
server_1         | During handling of the above exception, another exception occurred:
server_1         |
server_1         | Traceback (most recent call last):
server_1         |   File "/usr/lib/python3.7/site-packages/elasticsearch/connection/http_urllib3.py", line 217, in perform_request
server_1         |     method, url, body, retries=Retry(False), headers=request_headers, **kw
server_1         |   File "/usr/lib/python3.7/site-packages/urllib3/connectionpool.py", line 720, in urlopen
server_1         |     method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
server_1         |   File "/usr/lib/python3.7/site-packages/urllib3/util/retry.py", line 376, in increment
server_1         |     raise six.reraise(type(error), error, _stacktrace)
server_1         |   File "/usr/lib/python3.7/site-packages/urllib3/packages/six.py", line 735, in reraise
server_1         |     raise value
server_1         |   File "/usr/lib/python3.7/site-packages/urllib3/connectionpool.py", line 672, in urlopen
server_1         |     chunked=chunked,
server_1         |   File "/usr/lib/python3.7/site-packages/urllib3/connectionpool.py", line 387, in _make_request
server_1         |     conn.request(method, url, **httplib_request_kw)
server_1         |   File "/usr/lib/python3.7/http/client.py", line 1244, in request
server_1         |     self._send_request(method, url, body, headers, encode_chunked)
server_1         |   File "/usr/lib/python3.7/http/client.py", line 1290, in _send_request
server_1         |     self.endheaders(body, encode_chunked=encode_chunked)
server_1         |   File "/usr/lib/python3.7/http/client.py", line 1239, in endheaders
server_1         |     self._send_output(message_body, encode_chunked=encode_chunked)
server_1         |   File "/usr/lib/python3.7/http/client.py", line 1026, in _send_output
server_1         |     self.send(msg)
server_1         |   File "/usr/lib/python3.7/http/client.py", line 966, in send
server_1         |     self.connect()
server_1         |   File "/usr/lib/python3.7/site-packages/urllib3/connection.py", line 184, in connect
server_1         |     conn = self._new_conn()
server_1         |   File "/usr/lib/python3.7/site-packages/urllib3/connection.py", line 169, in _new_conn
server_1         |     self, "Failed to establish a new connection: %s" % e
server_1         | urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7efe93444b10>: Failed to establish a new connection: [Errno 111] Connection refused
server_1         |
server_1         | During handling of the above exception, another exception occurred:
server_1         |
server_1         | Traceback (most recent call last):
server_1         |   File "/usr/bin/waitress-serve-3", line 11, in <module>
server_1         |     load_entry_point('waitress==1.1.0', 'console_scripts', 'waitress-serve')()
server_1         |   File "/usr/lib/python3.7/site-packages/waitress/runner.py", line 259, in run
server_1         |     app = resolve(module, obj_name)
server_1         |   File "/usr/lib/python3.7/site-packages/waitress/runner.py", line 201, in resolve
server_1         |     obj = __import__(module_name, fromlist=segments[:1])
server_1         |   File "/opt/app/szurubooru/facade.py", line 152, in <module>
server_1         |     app = create_app()  # pylint: disable=invalid-name
server_1         |   File "/opt/app/szurubooru/facade.py", line 134, in create_app
server_1         |     wait_for_status='yellow', request_timeout=30)
server_1         |   File "/usr/lib/python3.7/site-packages/elasticsearch/client/utils.py", line 84, in _wrapped
server_1         |     return func(*args, params=params, **kwargs)
server_1         |   File "/usr/lib/python3.7/site-packages/elasticsearch/client/cluster.py", line 43, in health
server_1         |     "GET", _make_path("_cluster", "health", index), params=params
server_1         |   File "/usr/lib/python3.7/site-packages/elasticsearch/transport.py", line 353, in perform_request
server_1         |     timeout=timeout,
server_1         |   File "/usr/lib/python3.7/site-packages/elasticsearch/connection/http_urllib3.py", line 229, in perform_request
server_1         |     raise ConnectionError("N/A", str(e), e)
server_1         | elasticsearch.exceptions.ConnectionError: ConnectionError(<urllib3.connection.HTTPConnection object at 0x7efe93444b10>: Failed to establish a new connection: [Errno 111] Connection refused) caused by: NewConnectionError(<urllib3.connection.HTTPConnection object at 0x7efe93444b10>: Failed to establish a new connection: [Errno 111] Connection refused)
szurubooru_server_1 exited with code 1
elasticsearch_1  | [2019-10-10T02:44:44,994][INFO ][o.e.b.BootstrapChecks    ] [HUlIir5] bound or publishing to a non-loopback address, enforcing bootstrap checks
elasticsearch_1  | [2019-10-10T02:44:48,068][INFO ][o.e.c.s.MasterService    ] [HUlIir5] zen-disco-elected-as-master ([0] nodes joined)[, ], reason: new_master {HUlIir5}{HUlIir5mRm63LW4vWJN6iA}{Iph0JoD1R0OrzLZyJWKUMw}{172.18.0.2}{172.18.0.2:9300}
elasticsearch_1  | [2019-10-10T02:44:48,074][INFO ][o.e.c.s.ClusterApplierService] [HUlIir5] new_master {HUlIir5}{HUlIir5mRm63LW4vWJN6iA}{Iph0JoD1R0OrzLZyJWKUMw}{172.18.0.2}{172.18.0.2:9300}, reason: apply cluster state (from master [master {HUlIir5}{HUlIir5mRm63LW4vWJN6iA}{Iph0JoD1R0OrzLZyJWKUMw}{172.18.0.2}{172.18.0.2:9300} committed version [1] source [zen-disco-elected-as-master ([0] nodes joined)[, ]]])
elasticsearch_1  | [2019-10-10T02:44:48,102][INFO ][o.e.g.GatewayService     ] [HUlIir5] recovered [0] indices into cluster_state
elasticsearch_1  | [2019-10-10T02:44:48,111][INFO ][o.e.h.n.Netty4HttpServerTransport] [HUlIir5] publish_address {172.18.0.2:9200}, bound_addresses {0.0.0.0:9200}
```

**Comment by sgsunder** [comment]
There is a wait/block function in the app's entrypoint which checks to see if ElasticSearch is accessible, which times out after 30 seconds.

```
server_1         |   File "/opt/app/szurubooru/facade.py", line 134, in create_app
server_1         |     wait_for_status='yellow', request_timeout=30)
```

It seems like on your system, it takes longer than 30 seconds to get ElasticSearch online.

You can circumvent this for now by starting ElasticSearch separately:

```
# start elasticsearch and postgres separately:
docker-compose up -d elasticsearch sql
# monitor logs until elasticsearch and sql are up:
docker-compose logs -f
# start everything else:
docker-compose up -d
```

Can you let me know how long it seems to take for your machine to get ElasticSearch up and running? I can increase the timeout value based on that.

**Comment by mparkms** [comment]
It doesn't seem like ElasticSearch takes more than 10-15 seconds to initialize, but the error still persists. Your suggestion to start ES separately does work, so I'll stick with that for now.

**Comment by sgsunder** [comment]
So the ElasticSearch logs has to show not just that the port is open, but that the cluster state is indicating 'yellow'. Does it take longer than 15 seconds to indicate yellow?

**Comment by sgsunder** [comment]
Should be fixed in 4f5ea9c 
Let me know if a 2 minute timeout isn't enough.
